arq/worker.py

```diff
@@ -316,14 +321,35 @@ async def _poll_iteration(self) -> None:
 
         await self.start_jobs(job_ids)
 
-        for t in self.tasks:
+        if self.abort_jobs:
+            await self._scan_abort_jobs()
+
+        for job_id, t in list(self.tasks.items()):
             if t.done():
-                self.tasks.remove(t)
+                del self.tasks[job_id]
                 # required to make sure errors in run_job get propagated
                 t.result()
+            elif job_id in self._aborting_tasks:
+                t.cancel()
 
         await self.heart_beat()
 
+    async def _scan_abort_jobs(self) -> None:
```

Comment:
I know this is an internal method, but I think we still need a docstring explaining what's going on. It's not at all trivial.

In particular I'm not clear what `orphaned_job_keys` is or why it could happen.

Reply:
if `orphaned_job_keys` is just about clearing up keys added via `abort()` that don't apply here, I think cleaner to just omit it, those keys should expire soon anyway.